src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,534 @@
+/*
+ * Copyright (c) 2017, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+/**
+ * \file circuitpadding.h
+ * \brief Header file for circuitpadding.c.
+ **/
+#ifndef TOR_CIRCUITPADDING_H
+#define TOR_CIRCUITPADDING_H
+
+#include "circpad_negotiation.h"
+#include "lib/container/handles.h"
+#include "lib/evloop/timers.h"
+
+typedef struct circuit_t circuit_t;
+typedef struct origin_circuit_t origin_circuit_t;
+typedef struct cell_t cell_t;
+
+/**
+ * Circpad state specifier.
+ *
+ * Each circuit has up to two state machines, and each state
+ * machine consists of these states. Machines transition between
+ * these states using the event transition specifiers below.
+ */
+typedef enum {
+  CIRCPAD_STATE_START = 0,
+  CIRCPAD_STATE_BURST = 1,
+  CIRCPAD_STATE_GAP = 2,
+  CIRCPAD_STATE_END = 3
+} circpad_statenum_t;
+#define CIRCPAD_NUM_STATES  ((uint8_t)CIRCPAD_STATE_END+1)
+
+/**
+ * These constants form a bitfield to specify the types of events
+ * that can cause transitions between state machine states.
+ *
+ * Note that SENT and RECV are relative to this endpoint. For
+ * relays, SENT means packets destined towards the client and
+ * RECV means packets destined towards the relay. On the client,
+ * SENT means packets destined towards the relay, where as RECV
+ * means packets destined towards the client.
+ */
+typedef enum {
+  CIRCPAD_EVENT_NONPADDING_RECV = 1<<0,
+  CIRCPAD_EVENT_NONPADDING_SENT = 1<<1,
+  CIRCPAD_EVENT_PADDING_SENT = 1<<2,
+  CIRCPAD_EVENT_PADDING_RECV = 1<<3,
+  CIRCPAD_EVENT_INFINITY = 1<<4,
+  CIRCPAD_EVENT_BINS_EMPTY = 1<<5,
+  CIRCPAD_EVENT_LENGTH_COUNT = 1<<6
+} circpad_event_t;
+
+/** Boolean type that says if we decided to transition states or not */
+typedef enum {
+  CIRCPAD_STATE_UNCHANGED = 0,
+  CIRCPAD_STATE_CHANGED = 1
+} circpad_decision_t;
+
+/** The type for histogram bins */
+typedef uint16_t circpad_hist_bin_t;
+
+/** The type for absolute time, from monotime_absolute_usec() */
+typedef uint64_t circpad_time_t;
+
+/** The type for timer delays, in microseconds */
+typedef uint32_t circpad_delay_t;
+
+/**
+ * An infinite padding cell delay means don't schedule any padding --
+ * simply wait until a different event triggers a transition.
+ *
+ * This means that the maximum delay we can scedule is UINT32_MAX-1
+ * microseconds, or about 4300 seconds (1.25 hours).
+ * XXX: Is this enough if we want to simulate light, intermittent
+ * activity on an onion service?
```

Comment:
Hm, good point. This might be an issue in the future if we want to do simulate some very light activity.

What would happen if we used `uint64_t` for `circpad_delay_t`? How much more overhead for the structures?